python/pyspinw/plotting.py

```diff
@@ -0,0 +1,475 @@
+                        icell += 1
+            # keep unique within some tolerance (1e-3)
+            _, unique_idx = np.unique(np.round(dr,3), axis=0, return_index=True)
+            dr = dr[unique_idx]
+            # sort and get n shortest
+            isort = np.argsort(np.linalg.norm(dr, axis=1))
+            dr = dr[isort[:self.polyhedra_args.n_nearest]]
+            atom1_pos_xyz = self.transform_points_abc_to_xyz(atom1_pos_rlu)
+            verts_xyz = dr + atom1_pos_xyz
+            rank = np.linalg.matrix_rank(dr)
+            if rank == 3:
+                hull = ConvexHull(verts_xyz)
+                polyhedra.append(PolyhedronMesh(vertices=verts_xyz[hull.vertices], faces=hull.simplices))
+            elif rank == 2:
+                # transform to basis of polygon plane
+                *_, _, evecs_inv = np.linalg.svd(verts_xyz - verts_xyz[0])  # sorted in decreasing order of singular value
+                verts_2d = (evecs_inv @ verts_xyz.T).T
+                hull = ConvexHull(verts_2d[:, :-1]) # exclude last col (out of polygon plane - all have same value)
+                verts_xyz = np.vstack((atom1_pos_xyz, verts_xyz[hull.vertices], ))  # include central atom1 position as vertex
+                faces = self._label_2D_mesh_faces(verts_xyz)
+                polyhedra.append(PolyhedronMesh(vertices=verts_xyz, faces=faces))
+            else:
+                print('Polyhedron vertices must not be a line or point')
+        return polyhedra
+    
+    def _label_2D_mesh_faces(self, verts):
+        # assume centre point has index 0
+        nverts = verts.shape[0]
+        faces = np.zeros((nverts-1, 3), dtype=int)
+        faces[:,1] = np.arange(1,nverts)
+        faces[:,2] = np.arange(2,nverts + 1)
+        faces[-1,2] = 1  # close the shape by returning to first non-central vertex
+        return faces
+    
+    def _remove_vertices_outside_extent(self, verts):
+        # DO THIS BEFORE CONVERTING TO XYZ
+        # remove pairs of verts that correpsond to a line outside the extent
+        _, iremove = self._remove_points_outside_extent(verts)
+        iatom2 = (iremove % 2).astype(bool) # end point of pair of vertices
+        # for atom2 type vertex we need to remove previous row (atom1 vertex)
+        # for atom1 type vertex we need to remove the subsequent row (atom2 vertex)
+        iremove = np.hstack((iremove, iremove[iatom2]-1, iremove[~iatom2]+1))
+        return np.delete(verts, iremove, axis=0), iremove
+
+    def _remove_points_outside_extent(self, points, tol=1e-10):
+        # DO THIS BEFORE CONVERTING TO XYZ
+        iremove = np.flatnonzero(np.logical_or(np.any(points < -tol, axis=1), np.any((points - self.extent)>tol, axis=1)))
+        return np.delete(points, iremove, axis=0), iremove
+
+class UnitCellSimple:
+    def __init__(self, atoms_list=[], bonds={}, origin=np.array([0,0,0]), basis_vec=np.eye(3)):
```

Comment:
Right... so Python has a ["default mutable argument"](https://docs.python-guide.org/writing/gotchas/), where a default empty list `[]` and dict `{}` argument doesn't actually mean that it's a default empty list or dict, rather:

> A new list is created once when the function is defined, and the same list is used in each successive call.

(How ... did ... I ... not ... know ... this !!!)

So basically every "empty" constructor calls in line 38 appends more atoms/bonds to the `UnitCellSimple`....

So it might be better to make these into `None` and use the ternary operator in the assignments below rather than leave them as empty lists/dicts...

Reply:
Wow, OK I didn't know this! I will make them `None` by default